components/scream/src/share/util/scream_kokkos_utils.hpp

```diff
@@ -117,61 +119,136 @@ struct ExeSpaceUtils<Kokkos::Cuda> {
    * Of the C concurrently running teams, which "slot" is open
    * for the given team.
    */
   template <typename MemberType>
   KOKKOS_INLINE_FUNCTION
-  int get_workspace_idx(const MemberType& /* team_member */) const
-  {
+  int get_workspace_idx(const MemberType& team_member) const
+  { return 0; }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  void release_workspace_idx(const MemberType& team_member, int ws_idx) const
+  { }
+};
+
+template <typename ExeSpace = Kokkos::DefaultExecutionSpace>
+class TeamUtils : public _TeamUtilsCommonBase<ExeSpace>
+{
+ public:
+  template <typename TeamPolicy>
+  TeamUtils(const TeamPolicy& policy) : _TeamUtilsCommonBase<ExeSpace>(policy)
+  { }
+};
+
+/*
+ * Specialization for OpenMP execution space
+ */
 #ifdef KOKKOS_ENABLE_OPENMP
-    return omp_get_thread_num() / _team_size;
-#else
-    return 0;
-#endif
-  }
+template <>
+class TeamUtils<Kokkos::OpenMP> : public _TeamUtilsCommonBase<Kokkos::OpenMP>
+{
+ public:
+  template <typename TeamPolicy>
+  TeamUtils(const TeamPolicy& policy) : _TeamUtilsCommonBase<Kokkos::OpenMP>(policy)
+  { }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& team_member) const
+  { return omp_get_thread_num() / _team_size; }
 };
+#endif
 
 /*
- * Specialization for Cuda execution space. On GPU, we expect all teams to run
- * at once.
+ * Specialization for Cuda execution space.
  */
 #ifdef KOKKOS_ENABLE_CUDA
 template <>
-class TeamUtils<Kokkos::Cuda>
+class TeamUtils<Kokkos::Cuda> : public _TeamUtilsCommonBase<Kokkos::Cuda>
 {
-  int _num_teams;
+  using Device = Kokkos::Device<Kokkos::Cuda, typename Kokkos::Cuda::memory_space>;
+  using flag_type = int; // this appears to be the smallest type that correctly handles atomic operations
+  using view_1d = typename KokkosTypes<Device>::view_1d<flag_type>;
 
-public:
+  bool    _need_ws_sharing; // true if there are more teams in the policy than can be run concurrently
+  view_1d _open_ws_slots;   // indexed by ws-idx, true if in current use, else false
+
+ public:
   template <typename TeamPolicy>
-  TeamUtils(const TeamPolicy& policy) { _num_teams = policy.league_size(); }
+  TeamUtils(const TeamPolicy& policy) :
+    _TeamUtilsCommonBase<Kokkos::Cuda>(policy),
+    _need_ws_sharing(policy.league_size() > _num_teams),
+    _open_ws_slots("open_ws_slots", _need_ws_sharing ? _num_teams : 0)
+  { }
 
-  int get_num_concurrent_teams() const { return _num_teams; }
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& team_member) const
+  {
+    if (!_need_ws_sharing) {
+      return team_member.league_rank();
+    }
+    else {
+      int ws_idx = 0;
+      Kokkos::single(Kokkos::PerTeam(team_member), [&] () {
+        ws_idx = team_member.league_rank() % _num_teams;
+        while (!Kokkos::atomic_compare_exchange_strong(&_open_ws_slots(ws_idx), (flag_type) 0, (flag_type)1)) {
+          // or random?
```

Comment:
Have you try to clock the avg wait time with this algo vs random? Does it even matter? Meaning, with an avg, say, p3 run, would this latency be noticeable?

Reply:
I need to test that, thanks for reminding me.